server/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -278,16 +279,32 @@ public final String path() {
     }
 
     public boolean hasContent() {
-        return contentLength() > 0;
+        return contentLength() > 0 || isStreamedContent();
     }
 
     public int contentLength() {
-        return httpRequest.content().length();
+        if (httpRequest.body().isFull()) {
+            return httpRequest.body().asFull().bytes().length();
+        } else {
+            return 0;
+        }
     }
 
     public BytesReference content() {
         this.contentConsumed = true;
-        return httpRequest.content();
+        if (httpRequest.body().isFull()) {
+            return httpRequest.body().asFull().bytes();
+        } else {
+            return BytesArray.EMPTY;
+        }
```

Comment:
It's possible to log with stream. You can wrap stream handlers one into another and do metering, logging, sniffing, even filtering. Might be not too friendly, pseudo-code:
```java
currentHandler = httpRequest.body().contentStream().handler();
loggingBuffer = new SomeByteBuffer(); 
httpRequest.body().contentStream().setHandler((chunk, isLast) -> {
  copyChunk = chunk.copy();
  loggingBuffer.add(copyChunk);
  currentHandler.onNext(chunk, isLast);
});
```

Reply:
Not sure how viable this would be with how auditing works today. The request payload is [logged](https://github.com/elastic/elasticsearch/blob/0ac7b65581fd40fd4d88991d7ac41717aa3af2fe/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java#L74) when the Rest handler is invoked. At this point, we simply don't have the full body.

Reply:
We can add a bit of logic into HttpAggregator, when logging is available do full aggregation. But it's a quite confusing side effect of enabling logging. Yet not breaking.